### PR TITLE
[SPARK-34296][SQL] AggregateWindowFunction frame should not always use UnboundedPreceding

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
```
CREATE TABLE test_unboundedpreceding(SELLER_ID INT) using parquet;

SELECT
  DENSE_RANK() OVER (
    ORDER BY SELLER_ID ROWS BETWEEN 10 PRECEDING
    AND CURRENT ROW
  ) AS SELLER_RANK
FROM
  test_unboundedpreceding
```
It will throw:

Error: Error running query: org.apache.spark.sql.AnalysisException: Window Frame specifiedwindowframe(RowFrame, -10, currentrow$()) must match the required frame specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$()); (state=,code=0)
Related code:
https://github.com/apache/spark/blob/cde697a479a2f67c6bc4281f39a1ab2ff6a9d17d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala#L514

PostgresSQL
```
postgres@d40c1bcc8f5a:~$ psql
psql (11.3 (Debian 11.3-1.pgdg90+1))
Type "help" for help.

postgres=# create table test_unboundedpreceding(SELLER_ID int);
CREATE TABLE
postgres=#     SELECT
postgres-#       DENSE_RANK() OVER (
postgres(#         ORDER BY SELLER_ID ROWS BETWEEN 10 PRECEDING
postgres(#         AND CURRENT ROW
postgres(#       ) AS SELLER_RANK
postgres-#     FROM
postgres-#       test_unboundedpreceding;
 seller_rank
-------------
(0 rows)
```

In PostgresSQL, for all AggregationWindowFunction, it has a default frame, but user can write all kind of frame and finally it just use default frame. So in current PR, We will replace user-defined frame to default frame in `ResolveWindowFrame `
### Why are the changes needed?
Keep same with PostgresSQL

### Does this PR introduce _any_ user-facing change?
After this PR, user can write all kind of window frame in RangeLike window function, such as :
```
SELECT DENSE_RANK() OVER (
  ORDER BY salary
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary ROWS BETWEEN 5 PRECEDING AND CURRENT ROW
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary RANGE BETWEEN 2000 PRECEDING AND 1000 FOLLOWING
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary RANGE BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
 )
FROM basic_pays;

SELECT DENSE_RANK() OVER (
  ORDER BY salary RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
 )
FROM basic_pays;
```
### How was this patch tested?
Added UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 1]
```diff
@@ -633,7 +636,9 @@ case class CumeDist() extends RowNumberLike with SizeBasedWindowFunction {
   group = "window_funcs")
 // scalastyle:on line.size.limit line.contains.tab
 case class NthValue(input: Expression, offset: Expression, ignoreNulls: Boolean)
-    extends AggregateWindowFunction with OffsetWindowFunction with ImplicitCastInputTypes {
+    extends SpecifiedFrameAggregateWindowFunction
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:639]: Why we need to extend `SpecifiedFrameAggregateWindowFunction` even though `frame` with `UnspecifiedFrame` is overridden in L649?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 2]
```diff
@@ -511,14 +511,17 @@ case class Lag(
...
   override lazy val mergeExpressions =
     throw QueryExecutionErrors.mergeUnsupportedByWindowFunctionError
 }
 
-abstract class RowNumberLike extends AggregateWindowFunction {
+abstract class SpecifiedFrameAggregateWindowFunction extends AggregateWindowFunction {
+  override val frame: WindowFrame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
+}
+
+abstract class RowNumberLike extends SpecifiedFrameAggregateWindowFunction {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:524]: Why we can still use `UnboundedPreceding` for row number-like window funcs?

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:524]: > Why we can still use `UnboundedPreceding` for row number-like window funcs?

Since I found that some row number like function will change behavior when directly remove default frame in AggregateWindowFunction,  So I didi this. I need to check more about the behavior between Spark SQL and PostgresSQL

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:639]: > Why we need to extend `SpecifiedFrameAggregateWindowFunction` even though `frame` with `UnspecifiedFrame` is overridden in L649?

Replied in https://github.com/apache/spark/pull/31402#discussion_r567526579

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:524]: okay, thanks for the investigation. I think it would be better to describe more in the PR description for making the issue clearer.

**Comment by AngersZhuuuu** [comment]
PostgresSQL
```
postgres=# create table testData(val int, val_long bigint, val_double double precision, val_date date, val_timestamp timestamp, cate text);
CREATE TABLE
postgres=#  insert into testData values (null, 1, 1.0, '2017-08-01', to_timestamp(1501545600), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, 1, 1.0, '2017-08-01', to_timestamp(1501545600), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, 2, 2.5, '2017-08-02', to_timestamp(1502000000), 'a');
INSERT 0 1
postgres=#  insert into testData values (2, 2147483650, 100.001, '2020-12-31', to_timestamp(1609372800), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, null, 1.0, '2017-08-01', to_timestamp(1501545600), 'b');
INSERT 0 1
postgres=#  insert into testData values (2, 3, 3.3, '2017-08-03', to_timestamp(1503000000), 'b');
INSERT 0 1
postgres=#  insert into testData values (3, 2147483650, 100.001, '2020-12-31', to_timestamp(1609372800), 'b');
INSERT 0 1
postgres=#  insert into testData values (null, null, null, null, null, null);
INSERT 0 1
postgres=#  insert into testData values (3, 1, 1.0, '2017-08-01', to_timestamp(1501545600), null);
INSERT 0 1
postgres=# 
postgres=# 
postgres=# 
postgres=#  SELECT val, cate,
postgres-# rank() OVER w AS rank,
postgres-# dense_rank() OVER w AS dense_rank,
postgres-# cume_dist() OVER w AS cume_dist,
postgres-# percent_rank() OVER w AS percent_rank,
postgres-# ntile(2) OVER w AS ntile,
postgres-# row_number() OVER w AS row_number,
postgres-# var_pop(val) OVER w AS var_pop,
postgres-# var_samp(val) OVER w AS var_samp,
postgres-# covar_pop(val, val_long) OVER w AS covar_pop,
postgres-# corr(val, val_long) OVER w AS corr,
postgres-# stddev_samp(val) OVER w AS stddev_samp,
postgres-# stddev_pop(val) OVER w AS stddev_pop
postgres-# FROM testData
postgres-# WINDOW w AS (PARTITION BY cate ORDER BY val ASC nulls first)
postgres-# ORDER BY cate nulls first, val nulls first;
 val | cate | rank | dense_rank |     cume_dist      |    percent_rank    | ntile | row_number |        var_pop         |        var_samp        |     covar_pop     | corr |      stddev_samp       |       stddev_pop       
-----+------+------+------------+--------------------+--------------------+-------+------------+------------------------+------------------------+-------------------+------+------------------------+------------------------
     |      |    1 |          1 |                0.5 |                  0 |     1 |          1 |                        |                        |                   |      |                        |                       
   3 |      |    2 |          2 |                  1 |                  1 |     2 |          2 |                      0 |                        |                 0 |      |                        |                      0
     | a    |    1 |          1 |               0.25 |                  0 |     1 |          1 |                        |                        |                   |      |                        |                       
   1 | a    |    2 |          2 |               0.75 | 0.3333333333333333 |     2 |          3 |                      0 |                      0 |                 0 |      |                      0 |                      0
   1 | a    |    2 |          2 |               0.75 | 0.3333333333333333 |     1 |          2 |                      0 |                      0 |                 0 |      |                      0 |                      0
   2 | a    |    4 |          3 |                  1 |                  1 |     2 |          4 | 0.22222222222222222222 | 0.33333333333333333333 | 477218588.5555555 |    1 | 0.57735026918962576451 | 0.47140452079103168293
   1 | b    |    1 |          1 | 0.3333333333333333 |                  0 |     1 |          1 |                      0 |                        |                   |      |                        |                      0
   2 | b    |    2 |          2 | 0.6666666666666666 |                0.5 |     1 |          2 | 0.25000000000000000000 | 0.50000000000000000000 |                 0 |      | 0.70710678118654752440 | 0.50000000000000000000
   3 | b    |    3 |          3 |                  1 |                  1 |     2 |          3 | 0.66666666666666666667 | 1.00000000000000000000 |      536870911.75 |    1 | 1.00000000000000000000 | 0.81649658092772603273
(9 rows)

postgres=# 
```

Spark:
 before this patch :
```
```

After this patch:
```
-- !query
SELECT val, cate,
rank() OVER w AS rank,
dense_rank() OVER w AS dense_rank,
cume_dist() OVER w AS cume_dist,
percent_rank() OVER w AS percent_rank,
ntile(2) OVER w AS ntile,
row_number() OVER w AS row_number,
var_pop(val) OVER w AS var_pop,
var_samp(val) OVER w AS var_samp,
covar_pop(val, val_long) OVER w AS covar_pop,
corr(val, val_long) OVER w AS corr,
stddev_samp(val) OVER w AS stddev_samp,
stddev_pop(val) OVER w AS stddev_pop
FROM testData
WINDOW w AS (PARTITION BY cate ORDER BY val ASC nulls first)
ORDER BY cate nulls first, val nulls first
-- !query schema
struct<val:int,cate:string,rank:int,dense_rank:int,cume_dist:double,percent_rank:double,ntile:int,row_number:int,var_pop:double,var_samp:double,covar_pop:double,corr:double,stddev_samp:double,stddev_pop:double>
-- !query output
NULL	NULL	1	1	0.5	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
3	NULL	2	2	1.0	1.0	2	2	0.0	NULL	0.0	NULL	NULL	0.0
NULL	a	1	1	0.25	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
1	a	2	2	0.75	0.3333333333333333	2	3	0.0	0.0	0.0	NULL	0.0	0.0
1	a	2	2	0.75	0.3333333333333333	2	3	0.0	0.0	0.0	NULL	0.0	0.0
2	a	4	3	1.0	1.0	2	4	0.22222222222222224	0.33333333333333337	4.772185885555555E8	1.0	0.5773502691896258	0.4714045207910317
1	b	1	1	0.3333333333333333	0.0	1	1	0.0	NULL	NULL	NULL	NULL	0.0
2	b	2	2	0.6666666666666666	0.5	1	2	0.25	0.5	0.0	NULL	0.7071067811865476	0.5
3	b	3	3	1.0	1.0	2	3	0.6666666666666666	1.0	5.3687091175E8	1.0	1.0	0.816496580927726
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: > Why we can still use `UnboundedPreceding` for row number-like window funcs?

PostgresSQL
```
postgres=# create table testData(val int, val_long bigint, val_double double precision, val_date date, val_timestamp timestamp, cate text);
CREATE TABLE
postgres=#  insert into testData values (null, 1, 1.0, '2017-08-01', to_timestamp(1501545600), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, 1, 1.0, '2017-08-01', to_timestamp(1501545600), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, 2, 2.5, '2017-08-02', to_timestamp(1502000000), 'a');
INSERT 0 1
postgres=#  insert into testData values (2, 2147483650, 100.001, '2020-12-31', to_timestamp(1609372800), 'a');
INSERT 0 1
postgres=#  insert into testData values (1, null, 1.0, '2017-08-01', to_timestamp(1501545600), 'b');
INSERT 0 1
postgres=#  insert into testData values (2, 3, 3.3, '2017-08-03', to_timestamp(1503000000), 'b');
INSERT 0 1
postgres=#  insert into testData values (3, 2147483650, 100.001, '2020-12-31', to_timestamp(1609372800), 'b');
INSERT 0 1
postgres=#  insert into testData values (null, null, null, null, null, null);
INSERT 0 1
postgres=#  insert into testData values (3, 1, 1.0, '2017-08-01', to_timestamp(1501545600), null);
INSERT 0 1
postgres=# 
postgres=# 
postgres=# 
postgres=#  SELECT val, cate,
postgres-# rank() OVER w AS rank,
postgres-# dense_rank() OVER w AS dense_rank,
postgres-# cume_dist() OVER w AS cume_dist,
postgres-# percent_rank() OVER w AS percent_rank,
postgres-# ntile(2) OVER w AS ntile,
postgres-# row_number() OVER w AS row_number,
postgres-# var_pop(val) OVER w AS var_pop,
postgres-# var_samp(val) OVER w AS var_samp,
postgres-# covar_pop(val, val_long) OVER w AS covar_pop,
postgres-# corr(val, val_long) OVER w AS corr,
postgres-# stddev_samp(val) OVER w AS stddev_samp,
postgres-# stddev_pop(val) OVER w AS stddev_pop
postgres-# FROM testData
postgres-# WINDOW w AS (PARTITION BY cate ORDER BY val ASC nulls first)
postgres-# ORDER BY cate nulls first, val nulls first;
 val | cate | rank | dense_rank |     cume_dist      |    percent_rank    | ntile | row_number |        var_pop         |        var_samp        |     covar_pop     | corr |      stddev_samp       |       stddev_pop       
-----+------+------+------------+--------------------+--------------------+-------+------------+------------------------+------------------------+-------------------+------+------------------------+------------------------
     |      |    1 |          1 |                0.5 |                  0 |     1 |          1 |                        |                        |                   |      |                        |                       
   3 |      |    2 |          2 |                  1 |                  1 |     2 |          2 |                      0 |                        |                 0 |      |                        |                      0
     | a    |    1 |          1 |               0.25 |                  0 |     1 |          1 |                        |                        |                   |      |                        |                       
   1 | a    |    2 |          2 |               0.75 | 0.3333333333333333 |     2 |          3 |                      0 |                      0 |                 0 |      |                      0 |                      0
   1 | a    |    2 |          2 |               0.75 | 0.3333333333333333 |     1 |          2 |                      0 |                      0 |                 0 |      |                      0 |                      0
   2 | a    |    4 |          3 |                  1 |                  1 |     2 |          4 | 0.22222222222222222222 | 0.33333333333333333333 | 477218588.5555555 |    1 | 0.57735026918962576451 | 0.47140452079103168293
   1 | b    |    1 |          1 | 0.3333333333333333 |                  0 |     1 |          1 |                      0 |                        |                   |      |                        |                      0
   2 | b    |    2 |          2 | 0.6666666666666666 |                0.5 |     1 |          2 | 0.25000000000000000000 | 0.50000000000000000000 |                 0 |      | 0.70710678118654752440 | 0.50000000000000000000
   3 | b    |    3 |          3 |                  1 |                  1 |     2 |          3 | 0.66666666666666666667 | 1.00000000000000000000 |      536870911.75 |    1 | 1.00000000000000000000 | 0.81649658092772603273
(9 rows)

postgres=# 
```

Spark:
With  [this change](https://github.com/apache/spark/pull/31402#discussion_r567518596):
```
SELECT val, cate,
rank() OVER w AS rank,
dense_rank() OVER w AS dense_rank,
cume_dist() OVER w AS cume_dist,
percent_rank() OVER w AS percent_rank,
ntile(2) OVER w AS ntile,
row_number() OVER w AS row_number,
var_pop(val) OVER w AS var_pop,
var_samp(val) OVER w AS var_samp,
covar_pop(val, val_long) OVER w AS covar_pop,
corr(val, val_long) OVER w AS corr,
stddev_samp(val) OVER w AS stddev_samp,
stddev_pop(val) OVER w AS stddev_pop
FROM testData
WINDOW w AS (PARTITION BY cate ORDER BY val ASC nulls first)
ORDER BY cate nulls first, val nulls first
-- !query schema
struct<val:int,cate:string,rank:int,dense_rank:int,cume_dist:double,percent_rank:double,ntile:int,row_number:int,var_pop:double,var_samp:double,covar_pop:double,corr:double,stddev_samp:double,stddev_pop:double>
-- !query output
NULL	NULL	1	1	0.5	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
3	NULL	2	2	1.0	1.0	2	2	0.0	NULL	0.0	NULL	NULL	0.0
NULL	a	1	1	0.25	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
1	a	2	2	0.75	0.3333333333333333	1	2	0.0	0.0	0.0	NULL	0.0	0.0
1	a	2	2	0.75	0.3333333333333333	2	3	0.0	0.0	0.0	NULL	0.0	0.0
2	a	4	3	1.0	1.0	2	4	0.22222222222222224	0.33333333333333337	4.772185885555555E8	1.0	0.5773502691896258	0.4714045207910317
1	b	1	1	0.3333333333333333	0.0	1	1	0.0	NULL	NULL	NULL	NULL	0.0
2	b	2	2	0.6666666666666666	0.5	1	2	0.25	0.5	0.0	NULL	0.7071067811865476	0.5
3	b	3	3	1.0	1.0	2	3	0.6666666666666666	1.0	5.3687091175E8	1.0	1.0	0.816496580927726
```

Without  [this change](https://github.com/apache/spark/pull/31402#discussion_r567518596):
```
-- !query
SELECT val, cate,
rank() OVER w AS rank,
dense_rank() OVER w AS dense_rank,
cume_dist() OVER w AS cume_dist,
percent_rank() OVER w AS percent_rank,
ntile(2) OVER w AS ntile,
row_number() OVER w AS row_number,
var_pop(val) OVER w AS var_pop,
var_samp(val) OVER w AS var_samp,
covar_pop(val, val_long) OVER w AS covar_pop,
corr(val, val_long) OVER w AS corr,
stddev_samp(val) OVER w AS stddev_samp,
stddev_pop(val) OVER w AS stddev_pop
FROM testData
WINDOW w AS (PARTITION BY cate ORDER BY val ASC nulls first)
ORDER BY cate nulls first, val nulls first
-- !query schema
struct<val:int,cate:string,rank:int,dense_rank:int,cume_dist:double,percent_rank:double,ntile:int,row_number:int,var_pop:double,var_samp:double,covar_pop:double,corr:double,stddev_samp:double,stddev_pop:double>
-- !query output
NULL	NULL	1	1	0.5	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
3	NULL	2	2	1.0	1.0	2	2	0.0	NULL	0.0	NULL	NULL	0.0
NULL	a	1	1	0.25	0.0	1	1	NULL	NULL	NULL	NULL	NULL	NULL
1	a	2	2	0.75	0.3333333333333333	2	3	0.0	0.0	0.0	NULL	0.0	0.0
1	a	2	2	0.75	0.3333333333333333	2	3	0.0	0.0	0.0	NULL	0.0	0.0
2	a	4	3	1.0	1.0	2	4	0.22222222222222224	0.33333333333333337	4.772185885555555E8	1.0	0.5773502691896258	0.4714045207910317
1	b	1	1	0.3333333333333333	0.0	1	1	0.0	NULL	NULL	NULL	NULL	0.0
2	b	2	2	0.6666666666666666	0.5	1	2	0.25	0.5	0.0	NULL	0.7071067811865476	0.5
3	b	3	3	1.0	1.0	2	3	0.6666666666666666	1.0	5.3687091175E8	1.0	1.0	0.816496580927726
```


We can see that RankLike WindowFunction need default frame.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:524]: ping @maropu @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 3]
```diff
@@ -511,14 +511,15 @@ case class Lag(
 
 abstract class AggregateWindowFunction extends DeclarativeAggregate with WindowFunction {
   self: Product =>
-  override val frame: WindowFrame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:514]: @beliefer does `AggregateWindowFunction` support all kinds of window frames?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:514]: Yes. `AggregateWindowFunction` support all frames.

File: sql/core/src/test/resources/sql-functions/sql-expression-schema.md [hunk 4]
```diff
@@ -205,7 +205,7 @@
...
 | org.apache.spark.sql.catalyst.expressions.ParseToTimestamp | to_timestamp | SELECT to_timestamp('2016-12-31 00:12:00') | struct<to_timestamp(2016-12-31 00:12:00):timestamp> |
 | org.apache.spark.sql.catalyst.expressions.ParseUrl | parse_url | SELECT parse_url('http://spark.apache.org/path?query=1', 'HOST') | struct<parse_url(http://spark.apache.org/path?query=1, HOST):string> |
-| org.apache.spark.sql.catalyst.expressions.PercentRank | percent_rank | SELECT a, b, percent_rank(b) OVER (PARTITION BY a ORDER BY b) FROM VALUES ('A1', 2), ('A1', 1), ('A2', 3), ('A1', 1) tab(a, b) | struct<a:string,b:int,PERCENT_RANK() OVER (PARTITION BY a ORDER BY b ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW):double> |
+| org.apache.spark.sql.catalyst.expressions.PercentRank | percent_rank | SELECT a, b, percent_rank(b) OVER (PARTITION BY a ORDER BY b) FROM VALUES ('A1', 2), ('A1', 1), ('A2', 3), ('A1', 1) tab(a, b) | struct<a:string,b:int,PERCENT_RANK() OVER (PARTITION BY a ORDER BY b ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW):double> |
```
Comment (beliefer) [inline sql/core/src/test/resources/sql-functions/sql-expression-schema.md:208]: Why we change `ROWS` -> `RANGE`  ?

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-functions/sql-expression-schema.md:208]: > Why we change `ROWS` -> `RANGE` ?

Since we remove `AggregationWindowFunction`'s default. then will add a default frame in  `ResolveWindowFrame`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 5]
```diff
@@ -528,6 +529,10 @@ abstract class RowNumberLike extends AggregateWindowFunction {
   override val updateExpressions: Seq[Expression] = rowNumber + one :: Nil
 }
 
+trait SpecifiedFrameAggregateWindowFunction extends AggregateWindowFunction {
+  override val frame: WindowFrame = SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
+}
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:534]: It seems we no need this trait. If we want window function support using specified window frame, we just do like this:
`override val frame: WindowFrame = UnspecifiedFrame`

File: sql/core/src/test/resources/sql-functions/sql-expression-schema.md (hunk 4 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-functions/sql-expression-schema.md:208]: I think not change the origin behavior unless you have the determined reason.

**Comment by cloud-fan** [comment]
> Does this PR introduce any user-facing change?
> No

Can we have more description there? This definitely has user-facing change, right? More queries are allowed than before.

**Comment by AngersZhuuuu** [comment]
> > Does this PR introduce any user-facing change?
> > No
> 
> Can we have more description there? This definitely has user-facing change, right? More queries are allowed than before.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -3170,7 +3170,9 @@ class Analyzer(override val catalogManager: CatalogManager)
         WindowExpression(wf, s.copy(frameSpecification = wf.frame))
       case we @ WindowExpression(e, s @ WindowSpecDefinition(_, o, UnspecifiedFrame))
           if e.resolved =>
-        val frame = if (o.nonEmpty) {
+        val frame = if (e.isInstanceOf[RankLike]) {
+          SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3174]: what happens if order spec is present? do we force to use row frame?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 7]
```diff
@@ -633,7 +633,9 @@ case class CumeDist() extends RowNumberLike with SizeBasedWindowFunction {
   group = "window_funcs")
 // scalastyle:on line.size.limit line.contains.tab
 case class NthValue(input: Expression, offset: Expression, ignoreNulls: Boolean)
-    extends AggregateWindowFunction with OffsetWindowFunction with ImplicitCastInputTypes {
+    extends AggregateWindowFunction
+    with OffsetWindowFunction
+    with ImplicitCastInputTypes {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:638]: unnecessary change.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: > unnecessary change.

Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > what happens if order spec is present? do we force to use row frame?

In PostgresSQL, for all AggregationWindowFunction, it has a default frame, but user can write all kind of frame and finally it just use default frame. So in current PR, We can replace user-defined frame to default frame in `ResolveWindowFrame`, how about current change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 8]
```diff
@@ -3162,6 +3162,9 @@ class Analyzer(override val catalogManager: CatalogManager)
       case WindowExpression(wf: FrameLessOffsetWindowFunction,
         WindowSpecDefinition(_, _, f: SpecifiedWindowFrame)) if wf.frame != f =>
         throw QueryCompilationErrors.cannotSpecifyWindowFrameError(wf.prettyName)
+      case WindowExpression(wf: AggregateWindowFunction, s: WindowSpecDefinition)
+        if wf.frame != UnspecifiedFrame =>
+        WindowExpression(wf, s.copy(frameSpecification = wf.frame))
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3167]: It seems duplicate with https://github.com/apache/spark/blob/a5af58bde544985d152bd974aa18cf0bdbfad2d5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3171

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3167]: > It seems duplicate with
> 
> https://github.com/apache/spark/blob/a5af58bde544985d152bd974aa18cf0bdbfad2d5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3171

Yea, but logically they are not same.
This behavior should be earlier before throw exception.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3167]: This block is to use the default frame regardless of the specified frame ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: It seems we should investigate the behavior of other database.

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Hive 2.3.7 don't allow to specify frame.
WechatIMG75

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 8 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3167]: > This block is to use the default frame regardless of the specified frame ?

Yea..PG do like this ==

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3167]: PG is not the golden standard Spark has to follow. To me, the PG behavior looks unreasonable as it silently ignores the user input. Can we investigate more databases? I'd expect that we should fail if a user specified window frame but the window function can only use its default frame.

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/results/window.sql.out [hunk 9]
```diff
@@ -1054,3 +1054,219 @@ b	5	NULL	x	y	z	x	z
...
+9
+
+
+-- !query
+SELECT DENSE_RANK() OVER (
+  ORDER BY salary ROWS BETWEEN 5 PRECEDING AND CURRENT ROW
+ )
+FROM basic_pays
+-- !query schema
+struct<DENSE_RANK() OVER (ORDER BY salary ASC NULLS FIRST ROWS BETWEEN 5 PRECEDING AND CURRENT ROW):int>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/window.sql.out:1119]: It seems to be that we return the wrong answer here. `5 PRECEDING` is not respected.

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/window.sql.out:1119]: Thus, I'm -1 to this change.

Comment (AngersZhuuuu) [reply sql/core/src/test/resources/sql-tests/results/window.sql.out:1119]: All right, PG's behavior look not reasonable. Close this PR
